src/cocotb/handle.py

```diff
@@ -939,67 +1145,130 @@ def _set_value(self, value, call_sim):
+            TypeError: If assignment is given a type other than :class:`int`.
 
-    def _set_value(self, value, call_sim):
-        """Set the value of the underlying simulation object to *value*.
+            OverflowError: If the value used in assignment is out of range of a 32-bit signed integer.
+        """
+        return self._handle.get_signal_val_long()
 
-        This operation will fail unless the handle refers to a modifiable
-        object, e.g. net, signal or variable.
+    @value.setter
+    def value(self, value: int) -> None:
+        self.set(value)
 
-        Args:
-            value (bytes): The value to drive onto the simulator object.
 
-        Raises:
-            TypeError: If target has an unsupported type for
-                 string value assignment.
+class StringObject(
+    IndexableValueObjectBase[bytes, bytes, IntegerObject],
+    ValueObjectBase[bytes, bytes],
+):
+    """A string simulation object.
 
-        .. versionchanged:: 1.4
-            Takes :class:`bytes` instead of :class:`str`.
-            Users are now expected to choose an encoding when using these objects.
+    This type is used when a ``string`` (VHDL or Verilog) simulation object is seen.
+    """
 
-        """
-        value, set_action = self._check_for_set_action(value)
+    def __init__(self, handle: simulator.gpi_sim_hdl, path: Optional[str]) -> None:
+        super().__init__(handle, path)
 
+    def _set_value(
+        self,
+        value: bytes,
+        action: _GPISetAction,
+        schedule_write: Callable[
+            [ValueObjectBase[Any, Any], Callable[..., None], Sequence[Any]], None
+        ],
+    ) -> None:
         if not isinstance(value, bytes):
             raise TypeError(
                 "Unsupported type for string value assignment: {} ({!r})".format(
                     type(value), value
                 )
             )
 
-        call_sim(self, self._handle.set_signal_val_str, set_action, value)
+        schedule_write(self, self._handle.set_signal_val_str, (action, value))
 
-    @ModifiableObject.value.getter
+    @property
     def value(self) -> bytes:
+        """The value of the simulation object.
+
+        :getter:
+            Returns the current value of the simulation object as a :class:`bytes`.
+
+        :setter:
+            Assigns a :class:`bytes` value at the end of the current delta cycle.
+            When the value's length is less than the simulation object's,
+            the value is padded with NUL (``'\0'``) characters up to the appropriate length.
+            When the value's length is greater than the simulation object's,
+            the value is truncated without a NUL terminator to the appropriate length,
+            without warning.
+
+        Strings in both Verilog and VHDL are byte arrays without any particular encoding.
+        Encoding must be done to turn Python strings into byte arrays.
+        Because there are many encodings, this step is left up to the user.
```

Comment:
Maybe link to https://docs.python.org/3/library/codecs.html#standard-encodings ?